CorfuStore: auto retry on SEQUENCER_OVERFLOW transaction aborts

hisundar opened the pull request:
## Overview

For high concurrency use cases where only new keys are getting inserted transactionally the current cache eviction (LFU) in sequencer server can cause transaction aborts which can be consumed and retried internally in CorfuStore.

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java:
```diff
@@ -172,15 +175,27 @@ public void commit() {
+                txBegin();
+                operations.forEach(Runnable::run);
+            } catch(TransactionAbortedException txAbort) {
+                if (txAbort.getAbortCause() == AbortCause.SEQUENCER_OVERFLOW) {
```
Looks like your implementation is retrying with the same timestamp, but in case of sequencer overflow, retrying with same timestamp would still fail right?

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java:
```diff
@@ -30,6 +32,7 @@
     private final String namespace;
     private Timestamp timestamp;
     private final List<Runnable> operations;
+    private static final int DEFAULT_NUM_TX_RETRIES = 3;
```
can you please make it not static? When it's not static we can make this parameter configurable, whet it's static it's hard coded

hisundar replied:
the parameter is already configurable at runtime via an explicit argument right?
This is an internal value if the explicit argument is not set.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java:
```diff
@@ -172,15 +175,27 @@ public void commit() {
-        } finally {
-            txEnd();
-            operations.clear();
+    public void commit(final Timestamp timestamp, int numRetries) {
```
if you make `DEFAULT_NUM_TX_RETRIES` non-static, then you don't need to pass  `numRetries` as a parameter.
And you can check that DEFAULT_NUM_TX_RETRIES is correct in TxBuilder's constructor. 
Which makes your code more resilient and secure

hisundar replied:
done. thanks

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java:
```diff
@@ -172,15 +175,27 @@ public void commit() {
+        if (numRetries <= 0) {
+            throw new IllegalArgumentException("TxBuilder::commit: numRetries <=0"+numRetries);
+        }
+        while (numRetries-- > 0) {
```
Please don't `shadow` numRetries variable. Use another variable, like `remainingRetries`

hisundar replied:
done. thanks

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java:
```diff
@@ -172,15 +175,27 @@ public void commit() {
+        if (numRetries <= 0) {
+            throw new IllegalArgumentException("TxBuilder::commit: numRetries <=0"+numRetries);
+        }
+        while (numRetries-- > 0) {
```
And also please don't reduce the variable and check the result at the same time. It's confusing and makes code more error prone

hisundar replied:
done. will use separate variable.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java:
```diff
@@ -172,15 +175,27 @@ public void commit() {
+                operations.forEach(Runnable::run);
+            } catch(TransactionAbortedException txAbort) {
+                if (txAbort.getAbortCause() == AbortCause.SEQUENCER_OVERFLOW) {
+                    continue;
```
Please add a record in the log that the tx can't be finished

hisundar replied:
done.

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java:
```diff
@@ -172,15 +175,27 @@ public void commit() {
+                }
+                throw txAbort;
+            } finally {
+                txEnd();
```
as I know `txEnd()` never used, we always use `txCommit()` or something like that. Please double-check

xnull commented on runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java:
```diff
@@ -172,15 +175,27 @@ public void commit() {
+            } finally {
+                txEnd();
+                operations.clear();
+                break;
```
if `finally` block is always get executed after you finish your operation, that means you will always exit after the first iteration and retry will never happen

hisundar replied:
thanks for the catch!

hisundar commented:
> Looks like your implementation is retrying with the same timestamp, but in case of sequencer overflow, retrying with same timestamp would still fail right?

Good point, I have modified it and added a check only for non-snapshot transactions.

annym commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651452&bid=15577763)

annym commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/collections/TxBuilder.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CorfuDB/CorfuDB/pullRequest?prid=4653424&bid=15584313)
